cranelift/codegen/src/isa/riscv64/inst/emit.rs

```diff
@@ -620,12 +620,28 @@ impl MachInstEmit for Inst {
                 let base = from.get_base_register();
                 let offset = from.get_offset_with_state(state);
                 let offset_imm12 = Imm12::maybe_from_i64(offset);
+                let label = from.get_label_with_sink(sink);
 
-                // TODO: We shouldn't just fall back to `LoadAddr` immediately. For `MachLabel`s
-                // we should try to emit the `auipc` and add a relocation on this load.
-                let (addr, imm12) = match (base, offset_imm12) {
+                let (addr, imm12) = match (base, offset_imm12, label) {
                     // If the offset fits into an imm12 we can directly encode it.
-                    (Some(base), Some(imm12)) => (base, imm12),
+                    (Some(base), Some(imm12), _) => (base, imm12),
+
+                    (_, _, Some(label)) => {
```

Comment:
These match clauses here make me pause, if a `label` is available this means that it's ignoring `offset` and `base`, right? Could this perhaps assert that the `offset` is 0 and the `base` is not present? (or update the match clause's pattern match). 

Also could the catch-all be updated to `(None, None, None)` and other patterns all asserted to not exist? 

My guess is that this probably works as-is with what the backend already generates, so I'm thinking that this might be a good place for a few extra assertions that no other shapes are generated.

Reply:
It wasn't until I stared at the implementations of `get_base_register` and `get_label_with_sink` that I realized exactly one of them will return `Some` for any address. This would be a little simpler maybe if there were one function that returned an enum with either a label or a register. I don't think that's necessary, but at least a comment might be nice; the updated pattern matching `Some(_)` made me scratch my head too.

Reply:
Does this latest commit make things slightly more readable? I'm not a big fan of this match, but I also don't want to directly match the AMode since I think that would be worse.

Reply:
I think the extra comments are very helpful, thanks!